src/traceme.c

```diff
@@ -277,49 +321,103 @@ tracer_thread(void *arg)
     attr.precise_ip = 3;
 
     // Acquire file descriptor through which to talk to Intel PT.
-    struct tracer_ctx *tr_ctx = (struct tracer_ctx *) arg;
-    tr_ctx->perf_fd = syscall(SYS_perf_event_open, &attr,
-        tr_ctx->target_pid, -1, -1, 0);
-    if (tr_ctx->perf_fd == -1) {
-        err(EXIT_FAILURE, "syscall");
+    ret = syscall(SYS_perf_event_open, &attr, target_pid, -1, -1, 0);
+
+clean:
+    if (fclose(pt_type_file) == -1) {
+        ret = -1;
     }
 
+    return ret;
+}
+
+/*
+ * Set up Intel PT buffers and start a poll() loop for reading out the trace.
+ *
+ * Thread exits with 0 on success and -1 on error.
+ */
+static void *
+tracer_thread(void *arg)
+{
+    DEBUG("tracer init");
+
+    int page_size = getpagesize();
+    struct tracer_thread_args *thr_args = (struct tracer_thread_args *) arg;
+    int base_size = (1 + thr_args->data_bufsize) * page_size;
+    int ret = 0, sem_posted = 0;
+    void *base = NULL;
+    void *aux = NULL;
+    struct perf_event_mmap_page *header = NULL;
+
     /*
      * Allocate buffers.
      *
      * Data buffer is preceded by one management page, hence `1 + data_bufsize'.
      */
-    int page_size = getpagesize();
-    void *base = mmap(NULL, (1 + tr_ctx->data_bufsize) * page_size, PROT_WRITE,
-        MAP_SHARED, tr_ctx->perf_fd, 0);
+    base = mmap(NULL, base_size, PROT_WRITE,
+        MAP_SHARED, thr_args->perf_fd, 0);
     if (base == MAP_FAILED) {
-        err(EXIT_FAILURE, "mmap");
+        ret = -1;
+        goto clean;
     }
 
-    struct perf_event_mmap_page *header = base;
+    header = base;
     void *data = base + header->data_offset;
     (void) data; // XXX We will need this in the future to detect packet loss.
     header->aux_offset = header->data_offset + header->data_size;
-    header->aux_size = tr_ctx->aux_bufsize * page_size;
-    tr_ctx->mmap_header = header;
+    header->aux_size = thr_args->aux_bufsize * page_size;
 
     // AUX mapped R/W so as to have a saturating ring buffer.
-    void *aux = mmap(NULL, header->aux_size, PROT_READ | PROT_WRITE,
-        MAP_SHARED, tr_ctx->perf_fd, header->aux_offset);
+    aux = mmap(NULL, header->aux_size, PROT_READ | PROT_WRITE,
+        MAP_SHARED, thr_args->perf_fd, header->aux_offset);
     if (aux == MAP_FAILED) {
-        err(EXIT_FAILURE, "mmap2");
+        ret = -1;
+        goto clean;
     }
-    tr_ctx->aux = aux;
 
-    // Resume the interpreter loop.
+    // Copy out arguments for the poll loop.
```

Comment:
I assume this comment should really be something like "copy out because soon we'll lose control over the struct and the values will be overwritten"?

Reply:
Yes, that's explained in the next comment down.

Reply:
It's a bit late by that point :)